catalog/solr/catalog-solr-offline-gazetteer/src/main/java/ddf/catalog/solr/offlinegazetteer/OfflineGazetteerPlugin.java

```diff
@@ -0,0 +1,206 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.solr.offlinegazetteer;
+
+import static ddf.catalog.Constants.SUGGESTION_BUILD_KEY;
+
+import ddf.catalog.data.Attribute;
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.types.Core;
+import ddf.catalog.data.types.Location;
+import ddf.catalog.operation.CreateResponse;
+import ddf.catalog.operation.DeleteResponse;
+import ddf.catalog.operation.QueryRequest;
+import ddf.catalog.operation.Update;
+import ddf.catalog.operation.UpdateResponse;
+import ddf.catalog.plugin.PluginExecutionException;
+import ddf.catalog.plugin.PostIngestPlugin;
+import ddf.catalog.plugin.PreQueryPlugin;
+import ddf.catalog.plugin.StopProcessingException;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.common.SolrInputDocument;
+import org.codice.ddf.spatial.geocoding.GeoEntryAttributes;
+import org.codice.solr.client.solrj.SolrClient;
+import org.codice.solr.factory.SolrClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OfflineGazetteerPlugin implements PostIngestPlugin, PreQueryPlugin {
+  private static final Logger LOGGER = LoggerFactory.getLogger(OfflineGazetteerPlugin.class);
+  private static final String GAZETTEER_METACARD_TAG = "gazetteer";
+  public static final String STANDALONE_GAZETTEER_CORE_NAME = "standalone-solr-gazetteer";
+
+  private final SolrClientFactory clientFactory;
+  private final SolrClient solrClient;
+
+  public OfflineGazetteerPlugin(SolrClientFactory clientFactory) {
+    this.clientFactory = clientFactory;
+    this.solrClient = clientFactory.newClient(STANDALONE_GAZETTEER_CORE_NAME);
+  }
+
+  @Override
+  public CreateResponse process(CreateResponse input) throws PluginExecutionException {
+    List<Metacard> gazetteerMetacards =
+        input
+            .getCreatedMetacards()
+            .stream()
+            .filter(this::isGazetteerMetacard)
+            .collect(Collectors.toList());
+
+    if (gazetteerMetacards.isEmpty()) {
+      return input;
+    }
+
+    try {
+      solrClient.add(
```

Comment:
❓ @pklinef @rzwiefel Okay so no async concerns but regarding the data itself, what if the first solr operation succeeds but the main catalog provider one fails? Can an admin go back in via CLI and delete things from the gazetteer solr core by ID? Or set of IDs? Or are those kinds of cases what the sync command is for?

Reply:
This is a `PostIngestPlugin` method so the Catalog provider has already succeeded by this point.  The worst case is that the add to the gazetteer collection would fail and would need to be added again using the sync command.  If something got into the gazetteer collection that should not be there then you would need to call the gazetteer removeall command before sync.

Reply:
Yeah even with ~70k records only takes 5-10 minutes to delete and resync all records. On a performant production system I'd guess you could make time on that as well.

Reply:
Cool 👍